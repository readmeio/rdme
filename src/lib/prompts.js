const semver = require('semver');
const { prompt } = require('enquirer');
const parse = require('parse-link-header');

exports.generatePrompts = (versionList, selectOnly = false) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to use an existing project version or create a new one?',
    skip() {
      return selectOnly;
    },
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return selectOnly ? false : this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return selectOnly ? true : this.enquirer.answers.option === 'update';
    },
    hint: '1.0.0',
  },
];

function specOptions(specList, parsedDocs, currPage, totalPages) {
  const specs = specList.map(s => {
    return {
      message: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs.prev.page) specs.push({ message: `< Prev (page ${currPage - 1} of ${totalPages})`, value: 'prev' });
  if (parsedDocs.next.page) {
    specs.push({ message: `Next (page ${currPage + 1} of ${totalPages}) >`, value: 'next' });
  }
  return specs;
}

const updateOasPrompt = (specList, parsedDocs, currPage, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async result(spec) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      return spec;
    },
  },
];

exports.createOasPrompt = (specList, parsedDocs, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to update an existing OAS file or create a new one?',
    choices: [
      { message: 'Update existing', value: 'update' },
      { message: 'Create a new spec', value: 'create' },
    ],
    async result(picked) {
      if (picked === 'update') {
        try {
          const { specId } = await prompt(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        } catch (e) {
          return null;
        }
      }
      return picked;
    },
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    initial: opts.newVersion || false,
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
    validate(val) {
      return semver.valid(semver.coerce(val)) ? true : this.styles.danger('Please specify a semantic version.');
    },
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || isUpdate?.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
