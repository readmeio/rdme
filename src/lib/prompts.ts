import type { Version } from '../cmds/versions';
import type { Response } from 'node-fetch';
import type { Choice, PromptObject } from 'prompts';

import parse from 'parse-link-header';
import semver from 'semver';

import promptTerminal from './promptWrapper';
import { handleRes } from './readmeAPIFetch';

interface Spec {
  _id: string;
  title: string;
}

type SpecList = Spec[];

interface ParsedDocs {
  next?: {
    page: number;
    url: string;
  };
  prev?: {
    page: number;
    url: string;
  };
}

function specOptions(specList: SpecList, parsedDocs: ParsedDocs, currPage: number, totalPages: number): Choice[] {
  const specs = specList.map(s => {
    return {
      description: `API Definition ID: ${s._id}`, // eslint-disable-line no-underscore-dangle
      title: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs?.prev?.page) {
    specs.push({
      description: 'Go to the previous page',
      title: `< Prev (page ${currPage - 1} of ${totalPages})`,
      value: 'prev',
    });
  }
  if (parsedDocs?.next?.page) {
    specs.push({
      description: 'Go to the next page',
      title: `Next (page ${currPage + 1} of ${totalPages}) >`,
      value: 'next',
    });
  }
  return specs;
}

const updateOasPrompt = (
  specList: SpecList,
  parsedDocs: ParsedDocs,
  currPage: number,
  totalPages: number,
  getSpecs: (url: string) => Promise<Response>,
): PromptObject[] => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async format(spec: string) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      } else if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      }

      return spec;
    },
  },
];

export function createOasPrompt(
  specList: SpecList,
  parsedDocs: ParsedDocs,
  totalPages: number,
  getSpecs: ((url: string) => Promise<Response>) | null,
): PromptObject[] {
  return [
    {
      type: 'select',
      name: 'option',
      message: 'Would you like to update an existing OAS file or create a new one?',
      choices: [
        { title: 'Update existing', value: 'update' },
        { title: 'Create a new spec', value: 'create' },
      ],
      async format(picked: 'update' | 'create') {
        if (picked === 'update') {
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        }

        return picked;
      },
    },
  ];
}

/**
 * Series of prompts to construct a version object,
 * used in our `versions:create` and `versions:update` commands
 */
export function versionPrompt(
  /** list of versions, used for prompt about which version to fork */
  versionList: Version[],
  /** existing version if we're performing an update */
  isUpdate?: {
    is_stable: boolean;
  },
): PromptObject[] {
  return [
    {
      // only runs for versions:create command
      type: isUpdate ? null : 'select',
      name: 'from',
      message: 'Which version would you like to fork from?',
      choices: versionList.map(v => {
        return {
          title: v.version,
          value: v.version,
        };
      }),
    },
    {
      // only runs for versions:update command
      type: !isUpdate ? null : 'text',
      name: 'newVersion',
      message: 'What should the version be renamed to?',
      hint: '1.0.0',
      validate(val: string) {
        // allow empty string, in which case the version won't be renamed
        if (!val) return true;
        return semver.valid(semver.coerce(val)) ? true : 'Please specify a semantic version.';
      },
    },
    {
      // if the existing version being updated is already the main version,
      // we can't switch that so we skip this question
      type: isUpdate?.is_stable ? null : 'confirm',
      name: 'is_stable',
      message: 'Would you like to make this version the main version for this project?',
    },
    {
      type: 'confirm',
      name: 'is_beta',
      message: 'Should this version be in beta?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be hidden.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_public',
      message: 'Would you like to make this version public?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be deprecated.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_deprecated',
      message: 'Would you like to deprecate this version?',
    },
  ];
}
