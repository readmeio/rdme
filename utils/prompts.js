exports.generatePrompts = versionList => [
  {
    type: 'select',
    name: 'option',
    message:
      "We couldn't find a version in ReadMe matching the version in your OAS file. Would you like to use an existing version or create a new one?",
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || (isUpdate && isUpdate.is_stable);
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
