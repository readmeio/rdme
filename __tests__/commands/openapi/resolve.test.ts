import fs from 'node:fs';

import prompts from 'prompts';
import { describe, beforeAll, beforeEach, afterEach, it, expect, vi, type MockInstance } from 'vitest';

import Command from '../../../src/commands/openapi/resolve.js';
import { runCommand, type OclifOutput } from '../../helpers/oclif.js';

describe('openapi resolve', () => {
  let fsWriteFileSyncSpy: MockInstance<typeof fs.writeFileSync>;
  let resolvedSpec: string;
  let run: (args?: string[]) => OclifOutput;
  let testWorkingDir: string;

  beforeAll(() => {
    run = runCommand(Command);
  });

  beforeEach(() => {
    testWorkingDir = process.cwd();

    fsWriteFileSyncSpy = vi.spyOn(fs, 'writeFileSync').mockImplementationOnce((filename, data) => {
      resolvedSpec = JSON.parse(data as string);
    });
  });

  afterEach(() => {
    process.chdir(testWorkingDir);
    vi.restoreAllMocks();
  });

  describe('resolving', () => {
    it.each([
      ['standard', 'circular-references.json'],
      ['recursive', 'recursive-reference.json'],
      ['recursive and nested circularity in a single file', 'combined-cases.json'],
      ['replace unprocessable circularity with empty objects', 'unresolvable-circular-references.json'],
      ['schema only contains $refs', 'circular-reference-ref-only.json'],
      ['schema only contains $refs with a `title`', 'circular-reference-ref-only-title.json'],
    ])('should support resolving circular references (case: %s)', async (_, specFile) => {
      const spec = require.resolve(`../../__fixtures__/circular-ref-oas/${specFile}`);
      const expectedOutput = require.resolve(
        `../../__fixtures__/circular-ref-oas/${specFile.replace('.json', '.resolved.json')}`,
      );

      prompts.inject(['output.json']);

      const result = await run([spec]);
      expect(result).toMatchSnapshot();

      expect(fsWriteFileSyncSpy).toHaveBeenCalledWith('output.json', expect.any(String));
      expect(resolvedSpec).toStrictEqual(JSON.parse(fs.readFileSync(expectedOutput, 'utf8')));
    });
  });

  describe('error handling', () => {
    it('should fail if supplied a file with no circular references', async () => {
      const spec = require.resolve('@readme/oas-examples/3.0/json/petstore.json');

      prompts.inject(['output.json']);

      const { error } = await run([spec]);
      expect(error).toStrictEqual(new Error('The file does not contain circular or recursive references.'));
    });

    it.each([['json'], ['yaml']])('should fail if given a Swagger 2.0 definition (format: %s)', async format => {
      const spec = require.resolve(`@readme/oas-examples/2.0/${format}/petstore.${format}`);

      await expect(run([spec])).resolves.toMatchObject({
        error: new Error('Sorry, this command only supports OpenAPI 3.0+ definitions.'),
      });
    });
  });
});
