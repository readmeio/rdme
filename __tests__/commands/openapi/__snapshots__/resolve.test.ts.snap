// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openapi resolve > resolving > should support resolving circular references (case: recursive and nested circularity in a single file) 1`] = `
{
  "result": {
    "outputPath": "output.json",
    "success": true,
  },
  "stderr": "- Validating the API definition located at /Users/erunion/code/readme/rdme/__tests__/__fixtures__/circular-ref-oas/combined-cases.json...
- Identifying and resolving circular/recursive references in your API definition...
✔ Identifying and resolving circular/recursive references in your API definition... done! ✅
",
  "stdout": "ℹ️  Unresolved circular references remain. These references will be replaced with empty objects.
Your API definition has been processed and saved to output.json!
",
}
`;

exports[`openapi resolve > resolving > should support resolving circular references (case: recursive) 1`] = `
{
  "result": {
    "outputPath": "output.json",
    "success": true,
  },
  "stderr": "- Validating the API definition located at /Users/erunion/code/readme/rdme/__tests__/__fixtures__/circular-ref-oas/recursive-reference.json...
- Identifying and resolving circular/recursive references in your API definition...
✔ Identifying and resolving circular/recursive references in your API definition... done! ✅
",
  "stdout": "Your API definition has been processed and saved to output.json!
",
}
`;

exports[`openapi resolve > resolving > should support resolving circular references (case: replace unprocessable circularity with empty objects) 1`] = `
{
  "result": {
    "outputPath": "output.json",
    "success": true,
  },
  "stderr": "- Validating the API definition located at /Users/erunion/code/readme/rdme/__tests__/__fixtures__/circular-ref-oas/unresolvable-circular-references.json...
- Identifying and resolving circular/recursive references in your API definition...
✔ Identifying and resolving circular/recursive references in your API definition... done! ✅
",
  "stdout": "Your API definition has been processed and saved to output.json!
",
}
`;

exports[`openapi resolve > resolving > should support resolving circular references (case: schema only contains $refs with a \`title\`) 1`] = `
{
  "result": {
    "outputPath": "output.json",
    "success": true,
  },
  "stderr": "- Validating the API definition located at /Users/erunion/code/readme/rdme/__tests__/__fixtures__/circular-ref-oas/circular-reference-ref-only-title.json...
- Identifying and resolving circular/recursive references in your API definition...
✔ Identifying and resolving circular/recursive references in your API definition... done! ✅
",
  "stdout": "Your API definition has been processed and saved to output.json!
",
}
`;

exports[`openapi resolve > resolving > should support resolving circular references (case: schema only contains $refs) 1`] = `
{
  "result": {
    "outputPath": "output.json",
    "success": true,
  },
  "stderr": "- Validating the API definition located at /Users/erunion/code/readme/rdme/__tests__/__fixtures__/circular-ref-oas/circular-reference-ref-only.json...
- Identifying and resolving circular/recursive references in your API definition...
✔ Identifying and resolving circular/recursive references in your API definition... done! ✅
",
  "stdout": "Your API definition has been processed and saved to output.json!
",
}
`;

exports[`openapi resolve > resolving > should support resolving circular references (case: standard) 1`] = `
{
  "result": {
    "outputPath": "output.json",
    "success": true,
  },
  "stderr": "- Validating the API definition located at /Users/erunion/code/readme/rdme/__tests__/__fixtures__/circular-ref-oas/circular-references.json...
- Identifying and resolving circular/recursive references in your API definition...
✔ Identifying and resolving circular/recursive references in your API definition... done! ✅
",
  "stdout": "Your API definition has been processed and saved to output.json!
",
}
`;
